RecoBTag/DeepBoostedJet/plugins/DeepBoostedJetTagsProducer.cc

```diff
@@ -0,0 +1,333 @@
+    psd0.add<std::vector<unsigned int>>("probTbq",       {3});
+    psd0.add<std::vector<unsigned int>>("probWcq",       {4});
+    psd0.add<std::vector<unsigned int>>("probWqq",       {5});
+    psd0.add<std::vector<unsigned int>>("probZbb",       {6});
+    psd0.add<std::vector<unsigned int>>("probZcc",       {7});
+    psd0.add<std::vector<unsigned int>>("probZqq",       {8});
+    psd0.add<std::vector<unsigned int>>("probHbb",       {9});
+    psd0.add<std::vector<unsigned int>>("probHcc",       {10});
+    psd0.add<std::vector<unsigned int>>("probHqqqq",     {11});
+    psd0.add<std::vector<unsigned int>>("probQCDbb",     {12});
+    psd0.add<std::vector<unsigned int>>("probQCDcc",     {13});
+    psd0.add<std::vector<unsigned int>>("probQCDb",      {14});
+    psd0.add<std::vector<unsigned int>>("probQCDc",      {15});
+    psd0.add<std::vector<unsigned int>>("probQCDothers", {16});
+    desc.add<edm::ParameterSetDescription>("flav_table", psd0);
+  }
+  desc.addOptionalUntracked<bool>("debugMode", false);
+
+  descriptions.add("pfDeepBoostedJetTags", desc);
+}
+
+std::unique_ptr<MXBufferFileCache> DeepBoostedJetTagsProducer::initializeGlobalCache(
+  const edm::ParameterSet& iConfig)
+{
+  // get the model files
+  std::string model_file = iConfig.getParameter<edm::FileInPath>("model_path").fullPath();
+  std::string param_file = iConfig.getParameter<edm::FileInPath>("param_path").fullPath();
+
+  // load the model and params and save it in the cache
+  MXBufferFileCache* cache = new MXBufferFileCache();
+  cache->model_data = new mxnet::BufferFile(model_file);
+  cache->param_data = new mxnet::BufferFile(param_file);
+  return std::unique_ptr<MXBufferFileCache>(cache);
+}
+
+void DeepBoostedJetTagsProducer::globalEndJob(const MXBufferFileCache* cache)
+{
+  if (cache->model_data != nullptr) {
+    delete cache->model_data;
+  }
+  if (cache->param_data != nullptr) {
+    delete cache->param_data;
+  }
+}
+
+void DeepBoostedJetTagsProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+
+  // initialize output collection
+  std::vector<std::unique_ptr<JetTagCollection>> output_tags;
+  for (std::size_t i=0; i < flav_pairs_.size(); i++) {
+    if (!tag_infos->empty()) {
+      auto jet_ref = tag_infos->begin()->jet();
+      output_tags.emplace_back(std::make_unique<JetTagCollection>(
+            edm::makeRefToBaseProdFrom(jet_ref, iEvent)));
+    } else {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>());
+    }
+  }
+
+  for (unsigned jet_n=0; jet_n<tag_infos->size(); ++jet_n){
+
+    const auto& taginfo = tag_infos->at(jet_n);
+    std::vector<float> outputs(flav_pairs_.size(), 0); // init as all zeros
+
+    if (!taginfo.features().empty()){
+      // convert inputs
+      make_inputs(taginfo);
+      // run prediction and get outputs
+      outputs = predictor_->predict(data_);
```

Comment:
When I ran a simplified version of this producer (https://github.com/TreeMaker/TreeMaker/blob/c3be78637ae6d4ba4692cd01b68a1b149f005a38/Utils/src/DeepAK8Producer.cc, using the [GitLab version](https://gitlab.cern.ch/DeepAK8/NNKit)) over 2018 prompt data (just for testing), I occasionally ran into `"Error running forward"` exceptions. This happened when I ran using 4 threads, but not when I reran the same event using 1 thread, so I suspect it is a data race or other thread-safety issue in the mxnet library.

The GitLab version used mxnet 1.1.0, while the version added to CMSSW is 1.2.0 (https://github.com/cms-sw/cmsdist/pull/4167). It is possible the data race was fixed in the newer version. However, this should be tested carefully.

Reply:
@kpedro88 This is very useful to know. When I was testing this PR, I tried running over the TTBar RelVal samples with 8 threads for 9000 events and did not see any error. Then, looking at the MXNet changelog from 1.1.0 to 1.2.0, I indeed noticed this one:
> Fixed race condition for CPUSharedStorageManager->Free and launched workers at iter init stage to avoid frequent relaunch (https://github.com/apache/incubator-mxnet/pull/10096). 

So I suspect this is the cause for what you saw, and moving to 1.2.0 should solve it. Of course, more tests and feedback are more than welcome :)

Reply:
It sometimes took more than 10K events before I saw an exception when running with 4 threads. @slava77 told me he might try to run it on KNL, in which case we will definitely find out if there are still data races in the 1.2.0 release of mxnet.

Reply:
OK, then I can probably try to run with more events. 

> @slava77 told me he might try to run it on KNL, in which case we will definitely find out if there are still data races in the 1.2.0 release of mxnet.

This would be very interesting to see :)